Update introduction to continuous aggregates (#517)
timescale/docs.timescale.com-content: using-timescaledb/continuous-aggregates.md

Aggregate queries which touch large swathes of time-series data can
take a long time to compute because the system needs to scan large
amounts of data on every query execution. To make such queries faster,

> continuous aggregates allows pre-computing (or materializing) the

Reviewer: People that are accustomed to databases are probably familiar with what materialization is, so not sure it is good to use "precomputing" instead of "materialization".

aggregates, while also providing means to continuously, and with low
overhead, keep them up-to-date as the underlying source data changes.

> Continuous aggregates are somewhat similar to PostgreSQL's
> [materialized views][postgres-materialized-views], but, unlike a
> materialized view, a continuous aggregate can be continuously and
> incrementally refreshed, either manually or via a policy that runs in
> the background. A refresh can cover the entire aggregate or just a
> specific time range. In either case, the refresh only recomputes the
> aggregate buckets that have changed since the last refresh.

Reviewer: That first sentence is really long and many nested phrases with commas between making it harder to read. I think you should split it up as it was in the original (you can replace the `;` with a `.`). Something like this:
```suggestion
Continuous aggregates are somewhat similar to PostgreSQL's
[materialized views][postgres-materialized-views] but, unlike a
materialized view, they do not need to be manually refreshed. A continuous aggregate can manually refreshed or refreshed incrementally using a policy that runs in
the background. A refresh can cover the entire aggregate or just a
specific time range. In either case, the refresh only recomputes the
aggregate buckets that have changed since the last refresh.
```

policy needs to be created to keep it refreshed.

You can create several continuous aggregates for the same
hypertable. For example, you could create another continuous aggregate

> view for daily data.

Reviewer: :+1: I really think we should use more third-person singular.

`WHERE`) even though it is possible to parallelize but we might add
support for this in a future version.

#### Automatic refresh with a continuous aggregate policy

Continuous aggregate policies can be configured to support different
use cases. For example, you might want to:

- have the continuous aggregate and the hypertable be in sync, even
  when data is removed from the hypertable, or
- keep the aggregate data in the continuous aggregate when removing
  source data from the hypertable.

These use cases are supported by different configuration to
[`add_continuous_aggregate_policy`][add-continuous-aggregate-policy].

This function takes three arguments:

- The parameter `start_offset` indicates the start of the refresh
  window relative to the current time when the policy executes.
- The parameter `end_offset` indicates the end of the refresh window
  relative to the current time when the policy executes.
- The parameter `schedule_interval` indicates the refresh interval in
  wall-clock time.

Similar to the `refresh_continuous_aggregate` function, providing
`NULL` to `start_offset` or `end_offset` makes the range open-ended

> and will extend to the beginning or end of time,
> respectively. However, it seldom makes sense to use `NULL` for the
> `end_offset`. Instead, it is recommended to set the `end_offset` such
> that at least one time bucket is excluded. For time-series data that
> see mostly in-order writes, the time buckets that still see lots of
> writes will quickly have out-of-date aggregates. Eliding the buckets
> will also provide better performance.

Reviewer: Slight rewrite. I do not think you need that last sentence though
```suggestion
and will extend to the beginning or end of time,
respectively. However, it seldom makes sense to use `NULL` for the
`end_offset`. Instead, it is recommended to set the `end_offset` so
that at least the last time bucket is excluded. Since time-series data usually
see mostly writes to the end of the table, any aggregate on this bucket will quickly become out-of-date. Eliding the last buckets
will also provide better performance.
```

Author: FWIW, I wanted to avoid the use of "last bucket" because it is a bit ambiguous and there is really no choice as to what buckets `end_offset` will exclude. I did some tweaks to the text, however.

Reviewer: Using "newest buckets" then? The text just mention "buckets", which makes it confusing to what bucket is actually referenced.

hypertable `conditions` and runs every hour, you would write:

```sql
SELECT add_continuous_aggregate_policy('conditions_summary_hourly',
	start_offset => NULL,
	end_offset => INTERVAL '1 h',
	schedule_interval => INTERVAL '1 h');
```

This will ensure that all data in the continuous aggregate is up to
date with the hypertable except the last hour and also ensure that we
do not try to refresh the last bucket of the continuous
aggregate. Since we give an open-ended `start_offset`, any data that
is removed from `conditions` (for example, by using `DELETE` or
[`drop_chunks`][api-drop-chunks]) will also be removed from
`conditions_summary_hourly`. In effect, the continuous aggregate will
always reflect the data in the underlying hypertable.

If you instead want to keep data in the continuous aggregate even if
the source data is removed from the underlying hypertable, you also
need to set the `start_offset` in way that is compatible with the
[data retention policy][sec-data-retention] on the source
hypertable. For example, if you have a retention policy that removes
data older than one month, you need to set `start_offset` to one month

> (or less) and thereby not refresh into to region of dropped data.

Reviewer 2: ```suggestion
(or less) and thereby not refresh the region of dropped data.
```

CALL refresh_continuous_aggregate('conditions_summary_hourly, NULL, NULL);
```

However, we do not recommend open-ended refreshes on continuous
aggregates when there is a continuous ingest of new data since that
would trigger a refresh of time buckets that are not yet completely
filled. It might also materialize a lot of data, increase write
amplification, and affect other policies such as data retention.

>:TIP: You should avoid refreshing time intervals that still see a lot
>of writes, which is usually the last bucket of the continuous
>aggregate. These intervals are still changing and are unlikely to
>produce accurate aggregates, while at the same time slowing down the
>ingest rate of the hypertable due to write amplification. If you want
>to include the latest bucket in your queries, you should instead rely
>on [real-time aggregation][real-time-aggregates].

The `schedule_interval` option to `add_continuous_aggregate_policy`
controls how frequently materialization jobs will be run. Setting a
shorter interval will refresh more frequently but at the same time
consume more background worker resources.

> #### Using `timescaledb.information` Views

Reviewer: Aren't there more of these in the document?

Author: I think so, but I changed this because the content there didn't fit within the current section which is about real-time aggregation. So, it needs to be its own subsection.

Reviewer: I'm only worried about consistency. Are there other sections that use `**whatever**` style header that need to be changed to `####whatever`?

CALL refresh_continuous_aggregate('conditions_summary_hourly, NULL, NULL);
```

However, we do not recommend open-ended refreshes on continuous
aggregates when there is a continuous ingest of new data since that
would trigger a refresh of time buckets that are not yet completely
filled. It might also materialize a lot of data, increase write
amplification, and affect other policies such as data retention.

>:TIP: You should avoid refreshing time intervals that still see a lot
>of writes, which is usually the last bucket of the continuous
>aggregate. These intervals are still changing and are unlikely to
>produce accurate aggregates, while at the same time slowing down the
>ingest rate of the hypertable due to write amplification. If you want
>to include the latest bucket in your queries, you should instead rely
>on [real-time aggregation][real-time-aggregates].

The `schedule_interval` option to `add_continuous_aggregate_policy`
controls how frequently materialization jobs will be run. Setting a
shorter interval will refresh more frequently but at the same time
consume more background worker resources.

> #### Using `timescaledb.information` Views####

Reviewer 2: ```suggestion
#### Using `timescaledb.information` Views
```
